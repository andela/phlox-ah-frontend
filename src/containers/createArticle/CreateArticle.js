import { connect } from 'react-redux';
import React, { Component } from 'react';
import IdleTimer from 'react-idle-timer';
import PropTypes from 'prop-types';
import { Row, Col, Input } from 'react-materialize';
import { WithContext as ReactTags } from 'react-tag-input';
import 'medium-editor/dist/css/medium-editor.css';
import 'medium-editor/dist/css/themes/default.css';

import './CreateArticle.scss';
import '../../styles/style.scss';
import { getAllTags } from '../../requests/TagRequests';
import { getAllCategory } from '../../requests/CategoryRequests';
import ArticleForm from '../../components/ArticleForm/ArticleForm';
import { TagObjectToString, convertIdToString } from '../../util/TagsHelper';
import { createArticle, updateArticle, publishArticle } from '../../requests/ArticleRequests';


/**
 * @class CreateArticle
 * @extends {Component}
 */
class CreateArticle extends Component {
  /**
   * @description - This method runs first in the class
   * @returns {object} articles
   * @memberof Articles
   */
  constructor() {
    super();

    this.state = {
      alertVisible: false,
      body: '',
      category: '',
      description: '',
      hasChanges: false,
      idleTimer: null,
      imgUrl: null,
      imageName: '',
      isCreated: false,
      tags: [],
      title: ''
    };

    this.create = this.create.bind(this);
    this.getAlertMessage = this.getAlertMessage.bind(this);
    this.handleAddition = this.handleAddition.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
    this.handleDrag = this.handleDrag.bind(this);
    this.handleEditorChange = this.handleEditorChange.bind(this);
    this.onImageChange = this.onImageChange.bind(this);
    this.onInputChange = this.onInputChange.bind(this);
    this.onPublishArticle = this.onPublishArticle.bind(this);
    this.save = this.save.bind(this);
    this.update = this.update.bind(this);
  }

  /**
    * @description - This method get runs before the component mount
    * @returns {object} void
    * @memberof Article
    */
  componentDidMount() {
    this.props.getAllCategory();
    this.props.getAllTags();
  }

  /**
    * @description - This method sets the title and description input values
    * @param {object} e
    * @returns {object} void
    * @memberof Article
    */
  onInputChange(e) {
    this.setState({ [e.target.id]: e.target.value, hasChanges: true, alertVisible: true });
  }

  /**
    * @description - This method sets the image
    * @param {object} e
    * @returns {object} void
    * @memberof Article
    */
  onImageChange(e) {
    const file = e.target.files[0];
    if (file.type === 'image/png' || file.type === 'image/jpg' || file.type === 'image/jpeg') {
      this.setState({
        imgUrl: file, imageName: file.name, hasChanges: true, alertVisible: true
      });
    }
  }

  /**
    * @description - This method sets the body input values
    * @param {objecj}  text
    * @returns {object} void
    * @memberof Article
    */
  handleEditorChange(text) {
    this.setState({ body: text, hasChanges: true, alertVisible: true });
  }

  /**
    * @description - This method show if the article has been save or not.
    * @returns {string} - return saved or not
    * @memberof Article
    */
  getAlertMessage() {
    if (this.state.hasChanges) {
      return 'Saving...';
    }
    return 'Saved';
  }

  /**
    * @description - This method set the article status to published.
    * @returns {string} - return saved or not
    * @memberof Article
    */
  onPublishArticle() {
    const status = 'published';
    const { slug } = this.props.article;
    const tags = TagObjectToString(this.state.tags);
    this.props.publishArticle({ slug, status, tags });
  }

  /**
    * @description - This method create the article if changes has been made
    * @returns {string} - return saved or not
    * @memberof Article
    */
  create() {
    const {
      hasChanges, title, description, body, category, imgUrl
    } = this.state;

    const categoryId = Number(category);

    const tags = TagObjectToString(this.state.tags);

    if (hasChanges) {
      this.setState({ hasChanges: false, alertVisible: true, isCreated: true });

      const formData = new FormData();
      formData.append('title', title);
      formData.append('description', description);
      formData.append('body', body);
      formData.append('imgUrl', imgUrl);
      formData.append('categoryId', categoryId);
      formData.append('tags', JSON.stringify(tags));
      this.props.createArticle(formData, tags);
    }
  }

  /**
    * @description - This method create or update the article if changes has been made
    * @returns {string} - return created or updated article
    * @memberof Article
    */
  save() {
    if (this.state.isCreated && !this.props.error) {
      this.update();
    } else {
      this.create();
    }
  }

  /**
    * @description - This method update the article if changes has been made
    * @returns {string} - return created or updated article
    * @memberof Article
    */
  update() {
    const {
      hasChanges, title, description, body, category, imgUrl
    } = this.state;

    const categoryId = Number(category);
    const tags = TagObjectToString(this.state.tags);
    const articleSlug = this.props.article.slug;

    if (hasChanges) {
      this.setState({ hasChanges: false, alertVisible: true });
      const formData = new FormData();
      formData.append('title', title);
      formData.append('description', description);
      formData.append('body', body);
      formData.append('imgUrl', imgUrl);
      formData.append('categoryId', categoryId);
      formData.append('tags', JSON.stringify(tags));

      this.props.updateArticle(formData, tags, articleSlug);
    }
  }

  /**
    * @description - This method delete the Tag input values
    * @param {objecj}  tagIndexToDelete
    * @returns {object} void
    * @memberof Article
    */
  handleDelete(tagIndexToDelete) {
    const { tags } = this.state;
    this.setState({
      tags: tags.filter((tag, index) => index !== tagIndexToDelete),
      hasChanges: true,
      alertVisible: true
    });
  }

  /**
    * @description - This method sets the Tag input values
    * @param {string}  tag
    * @returns {object} void
    * @memberof Article
    */
  handleAddition(tag) {
    this.setState(state => ({ tags: [...state.tags, tag], hasChanges: true, alertVisible: true }));
  }

  /**
    * @description - This method renders the category options value
    * @returns {object} - return object of category options
    * @memberof Article
    */
  renderSelectOptions() {
    return this.props.categories.map(
      category => <option key={category.id} value={category.id}>{category.category}</option>
    );
  }

  /**
    * @description - This method is use to drag the tag input values
    * @param {string} tag
    * @param {string} currentPosition
    *  @param {string} newPosition
    * @returns {object} void
    * @memberof Article
    */
  handleDrag(tag, currentPosition, newPosition) {
    const tags = [...this.state.tags];
    const newTags = tags.slice();

    newTags.splice(currentPosition, 1);
    newTags.splice(newPosition, 0, tag);
    this.setState({ tags: newTags });
  }

  /**
   * @description - This method renders the jsx for this component
   * @returns {jsx} - jsx
   * @memberof CreateArticle
   */
  render() {
    console.log(this.state);
    const defaultValue = this.state.category ? this.state.category : 0;
    return (
      <div>
        <IdleTimer
          ref={(ref) => { this.idleTimer = ref; }}
          timeout={7000}
          startOnMount={false}
          onIdle={this.save}>
          <Row className="create-article">
            <ArticleForm
              componentState={this.state}
              getAlertMessage={this.getAlertMessage()}
              onPublishArticle={this.onPublishArticle}
              onInputChange={this.onInputChange}
              handleEditorChange={this.handleEditorChange}
              onImageChange={this.onImageChange}
            />
            <Col m={4} l={3} className="tag-category">
              <label htmlFor="Title">Tags</label>
              <ReactTags
                tags={this.state.tags}
                suggestions={convertIdToString(this.props.suggestedTags)}
                handleDelete={this.handleDelete}
                handleAddition={this.handleAddition}
                handleDrag={this.handleDrag}
                required
              />
              <div className="input-field col s12">
                <span htmlFor="Title">Categories</span>
                <Input
                  s={12}
                  type="select"
                  id="category"
                  onChange={this.onInputChange}
                  defaultValue={defaultValue}
                  required>
                  <option value="0" disabled>Choose your option</option>
                  {this.renderSelectOptions()}
                </Input>
              </div>
            </Col>
          </Row>
        </IdleTimer>
      </div>
    );
  }
}

CreateArticle.propTypes = {
  article: PropTypes.object,
  categories: PropTypes.array,
  createArticle: PropTypes.func,
  error: PropTypes.bool,
  getAllCategory: PropTypes.func,
  getAllTags: PropTypes.func,
  publishArticle: PropTypes.func,
  suggestedTags: PropTypes.array,
  updateArticle: PropTypes.func
};

const mapStateToProps = (state) => {
  const {
    article, message, tags, loading, error
  } = state.article;
  const { categories } = state.category;
  const suggestedTags = state.tags.tags;

  return {
    article, message, tags, loading, categories, suggestedTags, error
  };
};

export default connect(mapStateToProps,
  {
    createArticle, updateArticle, publishArticle, getAllCategory, getAllTags
  })(CreateArticle);
