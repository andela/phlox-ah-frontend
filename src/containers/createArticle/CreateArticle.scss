$darkBlue: #0039cb;
$lightGray: #6d6d6d;
$darkGray: #424242;

body {
    font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif;
    font-size: 18px;
    line-height: 30px;
    background: #F8F9FA;
}

h1 {
    font-size: 60px;
    font-weight: bold;
    text-align: center;
    margin-bottom: 40px;
    padding-bottom: 40px;
    letter-spacing: -2px;
    border-bottom: 1px solid #dbdbdb;
}

h3 {
    font-size: 32px;
    line-height: 42px;
}

h4 {
    font-size: 26px;
    line-height: 32px;
}

p {
    margin-bottom: 20px;
}

a {
    color:black;
    text-decoration: underline;
    &:hover {
      color: $darkBlue;
    }
}

blockquote {
    display: block;
    padding-left: 20px;
    border-left: 6px solid $darkGray;
    margin-left: -15px;
    padding-left: 15px;
    font-style: italic;
    color: $lightGray;
}

.editable,
.secondEditable
 {
    outline: none;
    margin: 0 0 20px 0;
    padding: 0 0 20px 0;
}

/* Styles for selected tags */
div.ReactTags__tags {
  margin-top: 10px;
}

div.ReactTags__selected span.ReactTags__tag {
    background: #3B8CA1;
    color: #ffffff;
    font-size: 12px;
    display: inline-block;
    padding: 5px;
    margin: 5px 5px;
    cursor: move;
    border-radius: 2px;
}

div.react-tags-wrapper {
  background: #ffffff;
  padding: 10px 10px 0px;
}

div.ReactTags__selected a.ReactTags__remove {
    color: #ffffff;
    margin-left: 5px;
    cursor: pointer;
    text-decoration: none;
}

div.ReactTags__tagInput input.ReactTags__tagInputField {
  border: none;
}

/* Styles for suggestions */
div.ReactTags__suggestions {
    position: absolute;
    z-index: 90;
}

div.ReactTags__suggestions ul {
    list-style-type: none;
    box-shadow: .05em .01em .5em rgba(0,0,0,.2);
    background: white;
    width: 200px;
}

div.ReactTags__suggestions li {
    border-bottom: 1px solid #ddd;
    padding: 5px 10px;
    margin: 0;
}

div.ReactTags__suggestions li mark {
    text-decoration: underline;
    background: none;
    font-weight: 600;
}

div.ReactTags__suggestions ul li.ReactTags__activeSuggestion {
    background: #b7cfe0;
    cursor: pointer;
}

.status {
    font-size: 14px;
    margin-top: 6px;
    color: #777;
}

.flex {
  display: flex;
}

.create-article {
  margin:3px 30px;
  padding-top: 20px;
  background: #ffffff;
  padding-right: 30px;

  div.author-photo{
    width: 50px;
    height: 50px;
    background: #ccc;
    border-radius: 50%;
  }

  div.author-name {
    margin-left: 10px;
    font-size: 16px;
    line-height: 1.3;
    margin-top: 6px;
    color: #444;
  }

  div.pub-date {
    font-size: 12px;
    margin-top: 4px;
    color: #777;
  }

  .publish-btn-wrapper{
    text-align: right;
  }

  .publish-btn-wrapper button {
    border: 1px solid #0e718c;
    color: #0e718c;
    font-size: 15px;
    background-color: #fff;
    padding: 8px 6px;
    outline: none;
    border-radius: 3px;
    width: 130px;
    box-shadow: 1px 1px 0 #0e718c;
    cursor: pointer;
  }

  label.article-img-upload {
    position: relative;
    border: 1px solid #ddd;
    height: 40px;
    display: block;
    top: 39px;
    width: 100%;
    border-radius: 3px;
    min-width: 216px;

    p.file-name {
      position: absolute;
      top: 4px;
      margin: 0px;
      left: 12px;
      font-size: 15px;
      overflow: hidden;
      height: 30px;
    }

    label.browse-btn {
      position: absolute;
      right: 0px;
      font-size: 16px;
      border-radius: inherit;
      background: #ddd;
      padding: 4px 20px;
    }

    input.file-input{
      display: none;
    }
  }
}

.tag-category {
  background: #EEEEEE;
  margin-top: 90px;
  margin-left: 10px;
  height: 500px;
  padding-top: 10px;
  label, span {
    font-size: 16px;
    color: #5E5F63;
  };
  .input-field {
    margin-top: 40px;
  }
  .input-field .input-field {
    margin-top: 9px;
    background: #ffffff;
    padding: 10px 10px 0;
  }
}
