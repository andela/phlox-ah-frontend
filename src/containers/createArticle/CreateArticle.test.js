import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import CreateArticle from './CreateArticle';
import {
  article, categories, tags
} from '../../util/articleMockData';

const mockStore = configureMockStore();
const store = mockStore({
  article: {
    article,
    message: 'article created successfully',
    tags,
    loading: false,
    error: true
  },
  category: { categories },
  tags: { tags }
});
let component;

describe('<CreateArticle />', () => {
  beforeEach(() => {
    /* eslint-disable no-undef */
    const mockFunc = jest.fn();
    component = shallow(
      <CreateArticle
        store={store}
        createArticle={mockFunc}
        updateArticle={mockFunc}
        publishArticle={mockFunc}
        getAllCategory={mockFunc}
     />
    );
  });
  it('should render without throwing an error', () => {
    expect(component).toMatchSnapshot();
  });
});
