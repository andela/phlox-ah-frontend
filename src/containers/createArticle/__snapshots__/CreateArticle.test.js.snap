// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateArticle /> should render without throwing an error 1`] = `
<CreateArticle
  article={
    Object {
      "body": "<p>Writing can be better</p>",
      "categoryId": 2,
      "createdAt": "2018-11-09T08:05:04.173Z",
      "description": "writing skill",
      "id": 59,
      "imgUrl": "",
      "ratingAverage": 0,
      "readTime": 1,
      "slug": "writing-skill-b271c674-c842-4306-93ac-7597aa69824e",
      "status": "draft",
      "title": "writing skill",
      "updatedAt": "2018-11-09T08:05:04.173Z",
      "userId": 1,
    }
  }
  categories={
    Array [
      Object {
        "category": "religion",
        "createdAt": "2018-11-07T09:44:42.002Z",
        "id": 1,
        "updatedAt": "2018-11-07T09:44:42.002Z",
      },
    ]
  }
  createArticle={[Function]}
  error={true}
  getAllCategory={[Function]}
  getAllTags={[Function]}
  publishArticle={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  suggestedTags={
    Array [
      Object {
        "createdAt": "2018-11-07T09:39:44.428Z",
        "id": 1,
        "name": "five",
        "updatedAt": "2018-11-07T09:39:44.428Z",
      },
    ]
  }
  tags={
    Array [
      Object {
        "createdAt": "2018-11-07T09:39:44.428Z",
        "id": 1,
        "name": "five",
        "updatedAt": "2018-11-07T09:39:44.428Z",
      },
    ]
  }
  updateArticle={[Function]}
/>
`;
